controllers/replication.storage/volumegroupreplication_controller.go

```diff
@@ -33,21 +56,533 @@ type VolumeGroupReplicationReconciler struct {
+//+kubebuilder:rbac:groups=replication.storage.openshift.io,resources=volumegroupreplicationclasses,verbs=get;list
+//+kubebuilder:rbac:groups=replication.storage.openshift.io,resources=volumegroupreplicationcontents,verbs=get;list;watch;create;update;delete
+//+kubebuilder:rbac:groups=replication.storage.openshift.io,resources=volumereplications,verbs=get;list;watch;create;update;delete
+//+kubebuilder:rbac:groups=replication.storage.openshift.io,resources=volumereplications/status,verbs=get;list
+//+kubebuilder:rbac:groups=core,resources=persistentvolumeclaims;persistentvolumes,verbs=get;list;watch;update
+//+kubebuilder:rbac:groups=core,resources=persistentvolumeclaims/finalizers,verbs=update
 
 // Reconcile is part of the main kubernetes reconciliation loop which aims to
 // move the current state of the cluster closer to the desired state.
 func (r *VolumeGroupReplicationReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	logger := log.FromContext(ctx, "Request.Name", req.Name, "Request.Namespace", req.Namespace)
+
+	// Fetch VolumeGroupReplication instance
+	instance := &replicationv1alpha1.VolumeGroupReplication{}
+	err := r.Client.Get(ctx, req.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			logger.Info("volumeGroupReplication resource not found")
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	// Get VolumeGroupReplicationClass instance
+	vgrClassObj, err := r.getVolumeGroupReplicationClass(logger, instance.Spec.VolumeGroupReplicationClassName)
+	if err != nil {
+		logger.Error(err, "failed to get volumeGroupReplicationClass resource", "VGRClassName", instance.Spec.VolumeGroupReplicationClassName)
+		err := r.setGroupReplicationFailure(instance, logger, err)
+		return reconcile.Result{}, err
+	}
+
+	err = validatePrefixedParameters(vgrClassObj.Spec.Parameters)
+	if err != nil {
+		logger.Error(err, "failed to validate parameters of volumeGroupReplicationClass", "VGRClassName", instance.Spec.VolumeGroupReplicationClassName)
+		err := r.setGroupReplicationFailure(instance, logger, err)
+		return reconcile.Result{}, err
+	}
+
+	// Declare all dependent resources
+	vgrContentObj := &replicationv1alpha1.VolumeGroupReplicationContent{}
+	vrObj := &replicationv1alpha1.VolumeReplication{}
+
+	// Create/Update dependent resources only if the instance is not marked for deletion
+	if instance.GetDeletionTimestamp().IsZero() {
+
+		// Add finalizer to VGR instance
+		if err = AddFinalizerToVGR(r.Client, logger, instance); err != nil {
+			logger.Error(err, "failed to add VolumeGroupReplication finalizer")
+			return reconcile.Result{}, err
+		}
+
+		// Check if PVCs exist based on provided selectors
+		pvcList := corev1.PersistentVolumeClaimList{}
+		if instance.Spec.Source.Selector != nil {
+			labelSelector := ""
+			pvcList, labelSelector, err = r.getMatchingPVCsFromSource(instance, logger)
```

Comment:
add a check to ensure all pvc belongs to same driver/provisioner in the VolumeGroupReplicationClass.

Reply:
We are checking the driver details in the `getPVHandles` function at line: 575, though it doesn't error out if the proviosioner is different, but will update that.
I think that should do it?